File: typescript/base-hmi-types/src/converters/automation-xml/HmiScreenToAutomationXmlConverter.ts
import { HmiScreenBase } from "../../screens/base/HmiScreenBase.js";
import { getStaticValueOrDefault } from "../../screens/base/HmiProperty.js";
import { HmiScreenKind } from "../../screens/base/HmiScreenKind.js";
import { XmlWriter } from "./XmlWriter.js";

export class HmiScreenToAutomationXmlConverter {
  static writeAutomationXml(writer: XmlWriter, screen: HmiScreenBase, id: { value: number }): void {
    const elementName = getElementName(screen.kind);
    writer.writeStartElement(elementName);
    writer.writeAttributeString("ID", String(id.value++));
    writer.writeStartElement("AttributeList");
    writer.writeElementString("Name", screen.name);
    writer.writeElementString("Width", getStaticValueOrDefault(screen.width, 0));
    writer.writeElementString("Height", getStaticValueOrDefault(screen.height, 0));
    writer.writeEndElement("AttributeList");
    writer.writeEndElement(elementName);
  }
}

function getElementName(kind: HmiScreenKind): string {
  switch (kind) {
    case HmiScreenKind.Template:
      return "Hmi.Screen.ScreenTemplate";
    case HmiScreenKind.Popup:
      return "Hmi.Screen.ScreenPopup";
    case HmiScreenKind.Slidein:
      return "Hmi.Screen.ScreenSlidein";
    default:
      return "Hmi.Screen.Screen";
  }
}

File: src/BaseHmiTypes/Converters/AutomationXml/HmiScreenToAutomationXmlConverter.cs
using System.Globalization;
using System.Xml;
using BaseHmiTypes.Screens.Base;

namespace BaseHmiTypes.Converters.AutomationXml;

public static class HmiScreenToAutomationXmlConverter
{
    public static void WriteAutomationXml(XmlWriter writer, HmiScreenBase screen, ref int id)
    {
        writer.WriteStartElement(GetElementName(screen.Kind));
        writer.WriteAttributeString("ID", (id++).ToString(CultureInfo.InvariantCulture));
        writer.WriteStartElement("AttributeList");
        WriteElementString(writer, "Name", screen.Name);
        WriteElementString(writer, "Width", screen.Width.GetStaticValueOrDefault());
        WriteElementString(writer, "Height", screen.Height.GetStaticValueOrDefault());
        writer.WriteEndElement();

        writer.WriteEndElement();
    }

    private static string GetElementName(HmiScreenKind kind)
    {
        return kind switch
        {
            HmiScreenKind.Template => "Hmi.Screen.ScreenTemplate",
            HmiScreenKind.Popup => "Hmi.Screen.ScreenPopup",
            HmiScreenKind.Slidein => "Hmi.Screen.ScreenSlidein",
            _ => "Hmi.Screen.Screen"
        };
    }

    private static void WriteElementString(XmlWriter writer, string name, object? value)
    {
        writer.WriteElementString(name, value switch
        {
            null => string.Empty,
            bool boolean => boolean ? "true" : "false",
            IFormattable formattable => formattable.ToString(null, CultureInfo.InvariantCulture),
            _ => value.ToString()
        });
    }
}
